pkg/queue/priority_queue.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 Outreach Corporation. All Rights Reserved.
+//
+// Description: Priority queue.
+package queue
+
+import (
+	"container/heap"
+	"math"
+
+	"github.com/pkg/errors"
+)
+
+// const
+const (
+	// defPriorityQueueCapacity
+	defPriorityQueueCapacity = math.MaxInt32
+)
+
+// PriorityQueue provides priority queue feature.
+type PriorityQueue interface {
+	Push(item PriorityQueueItem) (PriorityQueueIndexItem, error)
+	Pop() PriorityQueueIndexItem
+	Peek() PriorityQueueIndexItem
+	Update(item PriorityQueueIndexItem)
+	Len() int
+}
+
+// PriorityQueueItem is required to be implemented for the item that is push
+// into the priority queue. Priority with smaller value has higher priority.
+type PriorityQueueItem interface {
+	GetData() interface{}
+	SetPriority(v int64)
+	GetPriority() int64
+}
+
+// PriorityQueueIndexItem provides additional access to the item index. It is used
+// by the priority queue internally.
+type PriorityQueueIndexItem interface {
+	PriorityQueueItem
+
+	// private method that is used by priority queue internally.
+	getIndex() int
+	setIndex(v int)
+}
+
+// PriorityQueueOption
+type PriorityQueueOption func(q *priorityQueue)
+
+// WithCapacity sets the queue capacity.
+func WithCapacity(v int) PriorityQueueOption {
+	return func(q *priorityQueue) {
+		q.capacity = v
+	}
+}
+
+// NewPriorityQueue creates a new priority queue.
+func NewPriorityQueue(opts ...PriorityQueueOption) PriorityQueue {
+	q := &priorityQueue{
+		capacity: defPriorityQueueCapacity,
+		queue:    new(priorityQueueInternal),
+	}
+	for _, opt := range opts {
+		opt(q)
+	}
+	if q.capacity <= 0 {
+		panic(errors.Errorf("invalid capacity: %v", q.capacity))
```

Comment:
Generally never panic in a library, return an error instead

Reply:
ok